File: build_stamp.py
"""Computes a content-hash "stamp" for the committed upload-page bundle.

The stamp lets the server (and the Mac `doctor` check) detect a stale
`dist/` build without re-running the front-end build: it hashes the exact
set of files that feed `yarn build`, so any edit to those inputs changes the
stamp, and `yarn build` writes the freshly computed stamp into
`dist/build-stamp.json` for later comparison.

This module is the Python half of a cross-language contract: the Node
build script `scripts/build-stamp.mjs` (added in a later task) must
reproduce STAMP_ALGORITHM byte-for-byte, so both sides agree on the stamp
for the same source tree without ever talking to each other.
"""

from __future__ import annotations

import hashlib
import json
from pathlib import Path, PurePosixPath

STAMP_ALGORITHM = (
    "build-stamp-v1: sha256 over sorted (relpath, sha256(LF-normalized "
    "bytes)) pairs for index.html, package.json, yarn.lock, tsconfig.json, "
    "vite.config.ts, vitest.config.ts, and everything under src/"
)

_TOP_LEVEL_INPUT_FILES = (
    "index.html",
    "package.json",
    "yarn.lock",
    "tsconfig.json",
    "vite.config.ts",
    "vitest.config.ts",
)


def _input_files(page_dir: Path) -> list[Path]:
    """Every build-input file under page_dir, per STAMP_ALGORITHM.

    Top-level files are included only if present (a fixture or a real page
    directory may be missing one); every file under src/ is included
    recursively. dist/, node_modules/, and dotfile directories are never
    walked into (src/ is the only directory walked, so this mostly matters
    for future-proofing if src/ ever nests one of those names).
    """
    files: list[Path] = []
    for name in _TOP_LEVEL_INPUT_FILES:
        candidate = page_dir / name
        if candidate.is_file():
            files.append(candidate)

    src_dir = page_dir / "src"
    if src_dir.is_dir():
        for path in src_dir.rglob("*"):
            if not path.is_file():
                continue
            if _is_excluded(path.relative_to(page_dir)):
                continue
            files.append(path)

    return files


def _is_excluded(relpath: Path) -> bool:
    parts = relpath.parts
    if "dist" in parts or "node_modules" in parts:
        return True
    # Any dotfile directory component (e.g. .git, .cache) excludes the file.
    # The file's own final component is not itself a directory, so it is
    # not checked here -- only the directories it lives under.
    return any(part.startswith(".") for part in parts[:-1])


def _posix_relpath(path: Path, page_dir: Path) -> str:
    return PurePosixPath(path.relative_to(page_dir).as_posix()).as_posix()


def _file_hash(path: Path) -> str:
    raw = path.read_bytes()
    normalized = raw.replace(b"\r\n", b"\n")
    return hashlib.sha256(normalized).hexdigest()


def compute_build_stamp(page_dir: Path) -> str:
    """sha256 hex stamp over the front-end build inputs under page_dir.

    See STAMP_ALGORITHM and the module docstring for the exact algorithm;
    this must match scripts/build-stamp.mjs byte-for-byte.
    """
    files = _input_files(page_dir)
    files.sort(key=lambda p: _posix_relpath(p, page_dir))

    manifest_parts: list[str] = []
    for path in files:
        relpath = _posix_relpath(path, page_dir)
        filehash = _file_hash(path)
        manifest_parts.append(f"{relpath}\n{filehash}\n")

    manifest = "".join(manifest_parts)
    return hashlib.sha256(manifest.encode("utf-8")).hexdigest()


def read_committed_stamp(page_dir: Path) -> str | None:
    """The `stamp` field of <page_dir>/dist/build-stamp.json, or None.

    Guarded against every way this can fail to exist or be well-formed: no
    dist/ dir, no build-stamp.json, unreadable file, invalid JSON, or a
    JSON value that isn't an object with a string `stamp` field.
    """
    stamp_path = page_dir / "dist" / "build-stamp.json"
    try:
        data = json.loads(stamp_path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None

    stamp = data.get("stamp") if isinstance(data, dict) else None
    return stamp if isinstance(stamp, str) else None


def bundle_is_current(page_dir: Path) -> bool:
    """True when the committed dist/ bundle's stamp matches its source."""
    committed = read_committed_stamp(page_dir)
    if committed is None:
        return False
    return committed == compute_build_stamp(page_dir)

File: upload_page/scripts/build-stamp.mjs
// Cross-language twin of build_stamp.py -- see that module's docstring and
// STAMP_ALGORITHM for the full spec. Both implementations must produce a
// byte-identical stamp for the same source tree without ever talking to
// each other; scripts/build-stamp.test.ts pins both sides to the same
// constant over a shared fixture tree.
import { createHash } from "node:crypto";
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  statSync,
  writeFileSync,
} from "node:fs";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const TOP_LEVEL_INPUT_FILES = [
  "index.html",
  "package.json",
  "yarn.lock",
  "tsconfig.json",
  "vite.config.ts",
  "vitest.config.ts",
];

/**
 * True when relPosixPath lives under a dist/, node_modules/, or dotfile
 * directory. Mirrors build_stamp.py's _is_excluded: only the directory
 * components are checked, never the file's own final name component.
 */
function isExcludedPath(relPosixPath) {
  const dirParts = relPosixPath.split("/").slice(0, -1);
  if (dirParts.includes("dist") || dirParts.includes("node_modules")) {
    return true;
  }
  return dirParts.some((part) => part.startsWith("."));
}

function toPosixRelative(pageDir, absPath) {
  return path.relative(pageDir, absPath).split(path.sep).join("/");
}

/** Collects every file under `dir`, recursively, as page-relative POSIX paths. */
function collectSrcFiles(dir, pageDir, out) {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const absPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      collectSrcFiles(absPath, pageDir, out);
    } else if (entry.isFile()) {
      const relPath = toPosixRelative(pageDir, absPath);
      if (!isExcludedPath(relPath)) {
        out.push(relPath);
      }
    }
  }
}

/** Every build-input file under pageDir, as page-relative POSIX paths. */
function inputFiles(pageDir) {
  const files = [];
  for (const name of TOP_LEVEL_INPUT_FILES) {
    const candidate = path.join(pageDir, name);
    if (existsSync(candidate) && statSync(candidate).isFile()) {
      files.push(name);
    }
  }

  const srcDir = path.join(pageDir, "src");
  if (existsSync(srcDir) && statSync(srcDir).isDirectory()) {
    collectSrcFiles(srcDir, pageDir, files);
  }

  return files;
}

/** sha256 hex of a file's bytes, with CRLF normalized to LF first. */
function fileHash(pageDir, relPosixPath) {
  const raw = readFileSync(path.join(pageDir, relPosixPath));
  // Byte-level CRLF -> LF, matching Python's raw.replace(b"\r\n", b"\n").
  // Round-tripping through the "binary" (latin1) encoding is lossless for
  // any byte value, so this stays a byte-for-byte replace, not a text one.
  const normalized = Buffer.from(
    raw.toString("binary").split("\r\n").join("\n"),
    "binary"
  );
  return createHash("sha256").update(normalized).digest("hex");
}

/**
 * sha256 hex stamp over the front-end build inputs under pageDir. Must
 * match build_stamp.py's compute_build_stamp byte-for-byte.
 */
export function computeBuildStamp(pageDir) {
  const relPaths = inputFiles(pageDir).sort();
  const manifest = relPaths
    .map((relPath) => `${relPath}\n${fileHash(pageDir, relPath)}\n`)
    .join("");
  return createHash("sha256").update(manifest, "utf-8").digest("hex");
}

function main() {
  const pageDir = path.resolve(fileURLToPath(import.meta.url), "..", "..");
  const stamp = computeBuildStamp(pageDir);
  const distDir = path.join(pageDir, "dist");
  mkdirSync(distDir, { recursive: true });
  writeFileSync(
    path.join(distDir, "build-stamp.json"),
    JSON.stringify({ stamp }) + "\n"
  );
  process.stdout.write(
    "build-stamp: wrote dist/build-stamp.json (" + stamp + ")\n"
  );
}

// ESM equivalent of Python's `if __name__ == "__main__":` -- only run main()
// when this file is the process entry point, not when a test imports it.
const isMainModule =
  process.argv[1] !== undefined &&
  import.meta.url === pathToFileURL(process.argv[1]).href;

if (isMainModule) {
  main();
}
